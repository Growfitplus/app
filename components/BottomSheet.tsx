import { Colors } from '@/constants/Colors';
import React, { FC, useEffect } from 'react';
import { Modal, StyleSheet, useWindowDimensions, View } from 'react-native';
import { Gesture, GestureDetector } from 'react-native-gesture-handler';
import Animated, { useAnimatedStyle, useSharedValue, withSpring } from 'react-native-reanimated';

interface BottomSheetProps {
  children: JSX.Element | JSX.Element[];
  isVisible: boolean;
  closeSheet: () => void;
}

const BottomSheet: FC<BottomSheetProps> = ({ children, isVisible, closeSheet }) => {
  const { height: screenHeight, width: screenWidth } = useWindowDimensions();
  const MAX_TRANSLATE_Y = 0;
  const MIN_TRANSLATE_Y = screenHeight * 0.05;
  const translateY = useSharedValue(0);
  const context = useSharedValue({ y: 0 });

  useEffect(() => {
    if (isVisible) {
      scrollTo(0);
    }
  }, [isVisible]);

  const gesture = Gesture.Pan()
    .onStart(() => {
      context.value = { y: translateY.value };
    })
    .onUpdate(e => {
      translateY.value = e.translationY + context.value.y;
      translateY.value = Math.max(translateY.value, MAX_TRANSLATE_Y);
    })
    .onEnd(() => {
      // Up -> 0
      // Down -> 20
      const movement = translateY.value;

      if (movement > MIN_TRANSLATE_Y) {
        translateY.value = withSpring(screenHeight);
        closeSheet();
      }
      if (movement < MIN_TRANSLATE_Y) {
        translateY.value = withSpring(MAX_TRANSLATE_Y);
      }
    });

  const scrollTo = (destination: number) => {
    'worklet';
    translateY.value = withSpring(destination, { damping: 50 });
  };

  const reanimatedBottomStyle = useAnimatedStyle(() => {
    return {
      transform: [{ translateY: translateY.value }],
    };
  });

  return (
    <Modal
      animationType='fade'
      transparent={true}
      visible={isVisible}
    >
      <View
        style={{
          alignItems: 'center',
          backgroundColor: 'rgba(0,0,0,0.5)',
          flex: 1,
          justifyContent: 'center',
        }}
      >
        <GestureDetector gesture={gesture}>
          <Animated.View style={[styles.main, reanimatedBottomStyle]}>
            <View
              style={{
                alignSelf: 'center',
                backgroundColor: '#000',
                borderRadius: 16,
                height: 4,
                width: screenWidth * 0.25,
              }}
            />
            {children}
          </Animated.View>
        </GestureDetector>
      </View>
    </Modal>
  );
};

const styles = StyleSheet.create({
  main: {
    backgroundColor: Colors.light['white+'],
    borderTopLeftRadius: 34,
    borderTopRightRadius: 34,
    bottom: 0,
    paddingHorizontal: 25,
    paddingTop: 16,
    position: 'absolute',
    width: '100%',
  },
});

export default BottomSheet;
