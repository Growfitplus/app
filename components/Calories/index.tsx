import React from 'react';
import { Platform, TextInput, View } from 'react-native';

import { useUserContext } from '@/contexts/user/context';
import { setCalories } from '@/contexts/user/actions';

import useToday from '@/hooks/useToday';

import { Colors } from '@/constants/Colors';
import { DefaultCaloriesGoal, MaxCalories } from '@/constants/Goals';
import { Fonts } from '@/constants/Fonts';

import { resetWeek } from '@/utils/resetWeek';

import CaloriesProgress from './CaloriesProgress';
import WeekDays from './WeekDays';
import { FireSVG } from '../SVG';
import Typography from '../Typography';
import NightTime from './NightTime';

const Calories = () => {
  const [
    {
      nutrition: { week, maintainWeight },
    },
    userDispatch,
  ] = useUserContext();
  const { today, hours } = useToday();
  const { calories } = week?.[today] || { calories: 0 };
  const caloriesGoal = maintainWeight ? maintainWeight : DefaultCaloriesGoal;
  const hasExceeded = calories > caloriesGoal;
  const showNight = hours >= 23 && hours < 7;

  const handleCalories = (value: string) => {
    if (Number(value) <= MaxCalories) {
      const updatedWeek = week.length === 0 ? resetWeek() : [...week];

      updatedWeek[today].calories = Number(value);
      updatedWeek[today].exceeded = Number(value) > caloriesGoal;
      updatedWeek[today].succeeded = Number(value) === caloriesGoal;

      userDispatch(setCalories(updatedWeek));
    }
  };

  return (
    <View
      style={{
        alignItems: 'center',
        backgroundColor: Colors.light['white+'],
        borderRadius: 24,
        paddingHorizontal: 42,
        paddingVertical: 36,
        position: 'relative',
        width: '100%',
      }}
    >
      {showNight && <NightTime />}
      <WeekDays />
      <View
        style={{
          alignItems: 'center',
          flexDirection: 'row',
          gap: 16,
          justifyContent: 'center',
          marginVertical: 32,
          width: '100%',
        }}
      >
        <FireSVG />
        <Typography
          weight='bold'
          customStyles={{ fontSize: 24 }}
        >
          Calorías
        </Typography>
      </View>
      <TextInput
        value={calories.toString()}
        onChangeText={handleCalories}
        inputMode='numeric'
        keyboardType={Platform.OS === 'ios' ? 'numbers-and-punctuation' : 'decimal-pad'}
        placeholder='0'
        style={{
          fontFamily: Fonts.RobotoMedium,
          fontSize: 56,
          marginBottom: 24,
          textAlign: 'center',
        }}
      />
      <CaloriesProgress />
      {calories < caloriesGoal && (
        <Typography
          customStyles={{
            color: Colors.light.gray[2],
            fontSize: 14,
          }}
        >
          Consume máximo {maintainWeight.toString()} Kcal
        </Typography>
      )}
      {hasExceeded && (
        <Typography
          customStyles={{
            color: Colors.light.gray[2],
            fontSize: 14,
          }}
        >
          {`Te excediste en ${calories - caloriesGoal} Kcal`}
        </Typography>
      )}
    </View>
  );
};

export default Calories;
