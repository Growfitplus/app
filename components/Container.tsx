import React from 'react';
import { StyleSheet, View, ViewStyle } from 'react-native';

interface ContainerProps {
  children: React.ReactNode;
  customStyles?: ViewStyle
}

const Container: React.FC<ContainerProps> = ({ children, customStyles }) => {
  return <View style={{...styles.container, ...customStyles}}>{children}</View>;
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    paddingHorizontal: 28
  }
})

export default Container;
