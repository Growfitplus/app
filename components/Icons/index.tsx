export { default as ArrowBackIcon } from './ArrowBackIcon';
export { default as ArrowNextIcon } from './ArrowNextIcon';
export { default as CameraIcon } from './CameraIcon';
export { default as ChevronNextIcon } from './ChevronNextIcon';
export { default as LogOutIcon } from './LogOutIcon';
