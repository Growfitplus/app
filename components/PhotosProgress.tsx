import React from 'react';
import { StyleSheet, useWindowDimensions, View } from 'react-native';
import { Image } from 'expo-image';

const PhotosProgress = ({ imageSource }: { imageSource: string }) => {
  const { width: widthScreen } = useWindowDimensions();

  const dimensions = Math.floor((widthScreen - 56) / 3) - 4

  return (
    <Image
      source={{ uri: imageSource }}
      style={{
        borderRadius: 16,
        width: dimensions,
        height: dimensions,
      }}
    />
  );
};

export default PhotosProgress;
