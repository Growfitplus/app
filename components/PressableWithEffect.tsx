import React from 'react';
import { Pressable, ViewStyle } from 'react-native';

interface PressableWithEffectProps {
  children: React.ReactNode;
  isEnabled?: boolean;
  customStyles?: ViewStyle;
  onPressAction?: () => void;
}

const PressableWithEffect: React.FC<PressableWithEffectProps> = ({
  children,
  customStyles,
  isEnabled = true,
  onPressAction,
}) => {
  return (
    <Pressable
      style={({ pressed }) => ({ ...customStyles, opacity: pressed ? 0.5 : 1 })}
      onPress={onPressAction}
      disabled={!isEnabled}
    >
      {children}
    </Pressable>
  );
};

export default PressableWithEffect;
