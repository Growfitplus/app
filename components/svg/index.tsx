export { default as ClockSVG } from './Clock';
export { default as DropSVG } from './Drop';
export { default as FireSVG } from './Fire';
export { default as HomeSVG } from './Home';
export { default as InfoSVG } from './Info';
export { default as UserSVG } from './User';
export { default as WalkSVG } from './Walk';
export { default as MoonSVG } from './Moon';
export { Logo as LogoSVG } from './Logo';
export { LogoLogIn as LogoLogInSVG } from './Logo';
