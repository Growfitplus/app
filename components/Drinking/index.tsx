import React from 'react';
import { View } from 'react-native';

import { useUserContext } from '@/contexts/user/context';
import { setLiters } from '@/contexts/user/actions';
import { useStorageContext } from '@/contexts/storage/context';

import { Colors } from '@/constants/Colors';
import { DrinkingGoal } from '@/constants/Goals';

import useToday from '@/hooks/useToday';
import { resetWeek } from '@/utils/resetWeek';

import Typography from '../Typography';
import { CheckIcon } from '../Icons';
import Drop from '../SVG/Drop';
import PressableWithEffect from '../PressableWithEffect';
import Add from '../SVG/Add';
import Subtract from '../SVG/Subtract';

const Drinking = () => {
  const [
    {
      nutrition: { week, litersGoal },
    },
    userDispatch,
  ] = useUserContext();
  const [{ isLoading }] = useStorageContext();
  const { today } = useToday();

  const { liters = 0 } = week[today];
  const goal = litersGoal ? litersGoal : DrinkingGoal;
  const percentage = liters < 3 ? Math.floor((liters / goal) * 100) : 100;

  const handleDrinking = (value: number) => {
    if (value >= 0 && value <= 6) {
      const updatedWeek = week.length === 0 ? resetWeek() : [...week];

      updatedWeek[today].liters = value;

      userDispatch(setLiters(updatedWeek));
    }
  };

  return (
    <View
      style={{
        backgroundColor: Colors.light['white+'],
        borderRadius: 24,
        flex: 1,
        paddingHorizontal: 24,
        paddingVertical: 36,
      }}
    >
      <View
        style={{
          alignItems: 'center',
          flexDirection: 'row',
          gap: 4,
        }}
      >
        <Drop />
        <Typography
          weight='bold'
          customStyles={{ fontSize: 12 }}
        >
          Agua
        </Typography>
      </View>
      <Typography customStyles={{ textAlign: 'center' }}>
        {percentage.toString()}% de tu meta
      </Typography>
      <View
        style={{
          alignItems: 'center',
          flexDirection: 'row',
          gap: 16,
          justifyContent: 'center',
          marginTop: 16,
          width: '100%',
        }}
      >
        <PressableWithEffect
          onPressAction={() => handleDrinking(liters - 1)}
          isEnabled={isLoading ? false : liters !== 0}
          handleDisableStyles={liters === 0}
        >
          <Subtract />
        </PressableWithEffect>
        <Typography
          weight='bold'
          customStyles={{
            color: liters > 0 ? Colors.light['dark+'] : Colors.light.gray[2],
            fontSize: 32,
            textAlign: 'center',
          }}
        >
          {liters}
        </Typography>
        <PressableWithEffect
          onPressAction={() => handleDrinking(liters + 1)}
          isEnabled={isLoading ? false : liters < 6}
          handleDisableStyles={liters === 6}
        >
          <Add />
        </PressableWithEffect>
      </View>
      <Typography
        customStyles={{
          color: Colors.light.gray[3],
          fontSize: 12,
          marginBottom: 16,
          textAlign: 'center',
        }}
      >
        litros
      </Typography>
      <View style={{ alignItems: 'center' }}>
        <View
          style={{
            alignItems: 'center',
            backgroundColor: liters >= 2 ? Colors.light['growfit+'] : 'transparent',
            borderRadius: 20,
            height: 20,
            justifyContent: 'center',
            width: 20,
          }}
        >
          <CheckIcon
            size={16}
            color={Colors.light.gray[1]}
          />
        </View>
      </View>
    </View>
  );
};

export default Drinking;
