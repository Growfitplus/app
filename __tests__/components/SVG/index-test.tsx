import { render } from '@testing-library/react-native';

import { ClockSVG, HomeSVG, InfoSVG, LogoLogInSVG, LogoSVG, UserSVG } from '@/components/SVG';

describe('SVG', () => {
  it('should render ClockSVG component', () => {
    const { toJSON } = render(<ClockSVG />);

    expect(toJSON).toMatchSnapshot();
  });

  it('should render HomeSVG component', () => {
    const { toJSON } = render(<HomeSVG />);

    expect(toJSON).toMatchSnapshot();
  });

  it('should render InfoSVG component', () => {
    const { toJSON } = render(<InfoSVG />);

    expect(toJSON).toMatchSnapshot();
  });

  it('should render LogoSVG component', () => {
    const { toJSON } = render(<LogoSVG />);

    expect(toJSON).toMatchSnapshot();
  });

  it('should render LogoLogInSVG component', () => {
    const { toJSON } = render(<LogoLogInSVG />);

    expect(toJSON).toMatchSnapshot();
  });

  it('should render UserSVG component', () => {
    const { toJSON } = render(<UserSVG />);

    expect(toJSON).toMatchSnapshot();
  });
});
