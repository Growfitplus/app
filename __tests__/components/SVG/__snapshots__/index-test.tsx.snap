// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SVG should render ClockSVG component 1`] = `[Function]`;

exports[`SVG should render HomeSVG component 1`] = `[Function]`;

exports[`SVG should render InfoSVG component 1`] = `[Function]`;

exports[`SVG should render LogoLogInSVG component 1`] = `[Function]`;

exports[`SVG should render LogoSVG component 1`] = `[Function]`;

exports[`SVG should render UserSVG component 1`] = `[Function]`;
